lsp-mode.el

```diff
@@ -3581,8 +3591,37 @@ A reference is highlighted only if it is visible in a window."
                 'def-params td-params)))
 
 (defun lsp--get-document-symbols ()
-  (lsp-request "textDocument/documentSymbol"
-               `(:textDocument ,(lsp--text-document-identifier))))
+  "Get document symbols.
+
+If the buffer has not been modified since symbols were last
+retrieved, simply return the latest result.
+
+Else, if the request was initiated by Imenu updating its menu-bar
+entry, perform it asynchronously; i.e., give Imenu the latest
+result and then force a refresh when a new one is available.
+
+Else (e.g., due to intereactive use of `imenu' or `xref'),
+perform the request synchronously."
+  (if (= (buffer-chars-modified-tick) lsp--document-symbols-tick)
+      lsp--document-symbols
+    (if (not lsp--document-symbols-request-async)
+	(setq lsp--document-symbols-tick (buffer-chars-modified-tick)
+	      lsp--document-symbols (lsp-request
+				     "textDocument/documentSymbol"
+				     `(:textDocument ,(lsp--text-document-identifier))))
+      (lsp-request-async "textDocument/documentSymbol"
+			 `(:textDocument ,(lsp--text-document-identifier)
+					 :mode alive)
+			 (lambda (document-symbols)
+			   (setq lsp--document-symbols-tick (buffer-chars-modified-tick)
```

Comment:
You should capture  (buffers-chars-modified-tick) before that since it could have changed after async call.